### PR TITLE
MOD-6750 Fix numeric range syntax

### PR DESCRIPTION
**Pull request by nafraf**
**Description**
This PR includes the following changes, which were included in parser v3 `DIALECT 5`:
- Allow signed params (also negative), e.g., `[-$param1 +$param1]`
- Allow signed exponent (also positive), e.g., `[1 20e+5]`
- Allow case sensitive infinity (also negative), e.g., `[-INF Inf]`

Additional changes:
- FILTER now also accepts `inf` (without sign) and not only `+inf` or `-inf`.
- FILTER now allows case sensitive infinity. e.g. `filter score -INF Inf`

This PR requires:
https://github.com/RediSearch/RediSearch/pull/4604 which creates the parser-v3

Not included in this PR:
- Disallow an exclusive (open ended) numeric, e.g., `[10 (((20] `
  (should only have a single leading open parenthesis for “exclusive” number)
  will be done in https://github.com/RediSearch/RediSearch/pull/4486
- Fail on range with 2 exclusive identical values
  [(10 (10] - Breaking - keep current behavior - it means a range with no values - can 
  [(10]- Breaking 
  will be done in https://github.com/RediSearch/RediSearch/pull/4486

**Which issues this PR fixes**
1. MOD-6750

**Main objects this PR modified**
1. parser v3

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/pytests/test_numbers.py [hunk 1]
```diff
@@ -389,3 +389,69 @@ def testNegativeValues(env):
...
+    env.assertEqual(res2, expected)
+
+    # Multiple parenthesis are allowed
+    expected = [1, 'doc05', ['n', '1.5e+2']]
+    res = env.cmd('FT.SEARCH', 'idx', '@n:[1e2 ((300]')
+    env.assertEqual(res, expected)
+    res = env.cmd('FT.SEARCH', 'idx', '@n:[((1 300]')
+    env.assertEqual(res, expected)
+
+    # invalid syntax - multiple signs are not allowed
```
Comment (oshadmi) [inline tests/pytests/test_numbers.py:455]: Test also multiple signs with `inf`, e.g., `--inf`

And also when using a parameter, e.g.,
`-$param` when param is `-inf` - should fail
`-$param` when param is `inf` - should become `-inf`?

File: tests/pytests/test_numbers.py [hunk 2]
```diff
@@ -389,3 +389,69 @@ def testNegativeValues(env):
...
+    env.assertEqual(conn.execute_command('HSET', 'doc13', 'n', '+iNf'), 1)
+    env.assertEqual(conn.execute_command('HSET', 'doc14', 'n', '-INF'), 1)
+    env.assertEqual(conn.execute_command('HSET', 'doc15', 'n', '-inf'), 1)
+    env.assertEqual(conn.execute_command('HSET', 'doc16', 'n', '-iNf'), 1)
+    env.assertEqual(conn.execute_command('HSET', 'doc17', 'n', '-1'), 1)
+
+    # Test unsigned numbers
+    expected = [3, 'doc01', ['n', '1.0'], 'doc02', ['n', '1'],
+                'doc03', ['n', '1.0e0']]
+    res = env.cmd('FT.SEARCH', 'idx', '@n:[1 1]')
```
Comment (oshadmi) [inline tests/pytests/test_numbers.py:419]: Consider simplifying using `NOCONTENT` (also in other occurrences)

File: tests/pytests/test_search_params.py [hunk 3]
```diff
@@ -325,15 +327,69 @@ def test_numeric_range(env):
...
     res2 = env.cmd('FT.SEARCH', 'idx', '@numval:[($min $max]', 'NOCONTENT', 'PARAMS', '4', 'min', '102', 'max', '+inf')
     env.assertEqual(res2, res1)
 
     res1 = env.cmd('FT.SEARCH', 'idx', '@numval:[-inf, (105]', 'NOCONTENT')
-    env.assertEqual(res1, [4, 'key1', 'key2', 'key3', 'key4'])
+    env.assertEqual(res1, [5, 'key1', 'key2', 'key3', 'key4', 'key6neg'])
     res2 = env.cmd('FT.SEARCH', 'idx', '@numval:[$min ($max]', 'NOCONTENT', 'PARAMS', '4', 'min', '-inf', 'max', '105')
     env.assertEqual(res2, res1)
 
+    res1 = env.cmd('FT.SEARCH', 'idx', '@numval:[-101 101]', 'NOCONTENT')
```
Comment (oshadmi) [inline tests/pytests/test_search_params.py:339]: Add comment on the block? (parameters with sign and/or exclusive ranges)

**Review by oshadmi — COMMENTED** [review]
Looks good 👍🏼 Few comments

File: tests/pytests/test_numbers.py (hunk 2 above)
Comment (nafraf) [reply tests/pytests/test_numbers.py]: OK. Done. Thanks

File: tests/pytests/test_numbers.py (hunk 1 above)
Comment (nafraf) [reply tests/pytests/test_numbers.py:451]: New tests added in `test_numbers.py` and `test_search_params.py`

I think that `-$param` when param is `-inf` - should be valid, because we are replacing the numeric value and after that applying the unary operator, it is the same case that `-$param` when param is `+3` it is valid and becomes `-3`

I modified the range limits validation of `+/-Inf` and the corresponding test in `test.py`, please check if it is correct.

File: tests/pytests/test_search_params.py (hunk 3 above)
Comment (nafraf) [reply tests/pytests/test_search_params.py:358]: OK. Done.

File: tests/pytests/test_numbers.py (hunk 1 above)
Comment (oshadmi) [reply tests/pytests/test_numbers.py:451]: Parameters would behave differently than plain values, for example, `--inf` fails but `-$param` succeeds when param is `-inf`, but since parameter substitutions occur in a different phase, it does make sense to support such valid parameter substitution use cases.

File: src/numeric_filter.c [hunk 4]
```diff
@@ -66,12 +73,12 @@ NumericFilter *NumericFilter_Parse(ArgsCursor *ac, QueryError *status) {
 
   // Parse the min range
   const char *s = AC_GetStringNC(ac, NULL);
-  if (parseDoubleRange(s, &nf->inclusiveMin, &nf->min, 1, status) != REDISMODULE_OK) {
+  if (parseDoubleRange(s, &nf->inclusiveMin, &nf->min, 1, 1, status) != REDISMODULE_OK) {
```
Comment (oshadmi) [inline src/numeric_filter.c:76]: Since legacy `FILTER` does not support parameters and sign is implicitly `+` ?

File: src/numeric_filter.c [hunk 5]
```diff
@@ -66,12 +73,12 @@ NumericFilter *NumericFilter_Parse(ArgsCursor *ac, QueryError *status) {
...
     NumericFilter_Free(nf);
     return NULL;
   }
   s = AC_GetStringNC(ac, NULL);
-  if (parseDoubleRange(s, &nf->inclusiveMax, &nf->max, 0, status) != REDISMODULE_OK) {
+  if (parseDoubleRange(s, &nf->inclusiveMax, &nf->max, 0, 1, status) != REDISMODULE_OK) {
```
Comment (oshadmi) [inline src/numeric_filter.c:81]: Since legacy `FILTER` does not support parameters and sign is implicitly `+` ?

**Review by oshadmi — COMMENTED** [review]
Few more comments

File: src/numeric_filter.c (hunk 4 above)
Comment (nafraf) [reply src/numeric_filter.c:76]: Yes. And a test was added to show that parameters are not supported.

File: src/numeric_filter.c (hunk 5 above)
Comment (nafraf) [reply src/numeric_filter.c:81]: Yes.

File: src/numeric_filter.c [hunk 6]
```diff
@@ -1,21 +1,25 @@
...
 #include "rmutil/vector.h"
 #include "query_param.h"
 
 int parseDoubleRange(const char *s, int *inclusive, double *target, int isMin,
-                            QueryError *status) {
-  if (isMin && !strcasecmp(s, "-inf")) {
+                      int sign, QueryError *status) {
+  if (isMin && (
+        (sign == 1 && !strcasecmp(s, "-inf")) ||
+        (sign == -1 && (!strcasecmp(s, "+inf") || !strcasecmp(s, "inf"))))) {
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/numeric_filter.c]: Done. Thanks.

File: src/numeric_filter.c [hunk 7]
```diff
@@ -1,21 +1,25 @@
...
     *target = NF_NEGATIVE_INFINITY;
     return REDISMODULE_OK;
-  } else if (!isMin && !strcasecmp(s, "+inf")) {
+  } else if (!isMin && (
+        (sign == 1 && (!strcasecmp(s, "+inf") || !strcasecmp(s, "inf"))) ||
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/numeric_filter.c]: Done. Thanks.

File: src/numeric_filter.c [hunk 8]
```diff
@@ -26,6 +30,9 @@ int parseDoubleRange(const char *s, int *inclusive, double *target, int isMin,
   char *endptr = NULL;
   errno = 0;
   *target = strtod(s, &endptr);
+  if(sign == -1) {
+    *target = -(*target);
+  }
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/numeric_filter.c]: OK.

File: tests/pytests/test.py [hunk 9]
```diff
@@ -1367,8 +1367,9 @@ def testNumericRange(env):
     env.expect('ft.create', 'idx', 'ON', 'HASH', 'schema', 'title', 'text', 'score', 'numeric', 'price', 'numeric').ok()
 
     env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', 5).error().contains("FILTER requires 3 arguments")
-    env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', 5, 'inf').error().contains("Bad upper range: inf")
+    env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', 5, '-inf').error().contains("Bad upper range: -inf")
```
[the remark this answers is not in the document]
Comment (nafraf) [reply tests/pytests/test.py:1370]: Tests added with `inf` (without sign) and using `inf` with uppercase letters.

**Review by oshadmi — DISMISSED** [review]
👍🏼 😎

File: tests/pytests/test.py [hunk 10]
```diff
@@ -1367,8 +1367,12 @@ def testNumericRange(env):
...
     env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', 'inf', 5).error().contains("Bad lower range: inf")
+    env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', '+inf', 5).error().contains("Bad lower range: +inf")
+    # Filter does not accept parameters
+    env.expect('ft.search', 'idx', 'hello kitty', 'filter', 'score', 5, '$n',
+               'PARAMS', 2, '10').error().contains("Bad upper range: $n")
```
Comment (raz-mon) [inline tests/pytests/test.py:1375]: I assume you meant to define the parameter `n` here?

File: tests/pytests/test.py [hunk 11]
```diff
@@ -1388,12 +1392,24 @@ def testNumericRange(env):
...
                                 "filter", "score", "(0", "(50")
-
         env.assertEqual(49, res[0])
+
         res = env.cmd('ft.search', 'idx', 'hello kitty', "nocontent",
                                 "filter", "score", "-inf", "+inf")
         env.assertEqual(100, res[0])
 
+        res = env.cmd('ft.search', 'idx', 'hello kitty', 'verbatim', "nocontent", "limit", 0, 100,
+                                "filter", "score", "-10", "(10")
```
Comment (raz-mon) [inline tests/pytests/test.py:1402]: Is this test related to this PR?

File: tests/pytests/test_numbers.py [hunk 12]
```diff
@@ -389,3 +389,91 @@ def testNegativeValues(env):
...
+    env.expect('FT.SEARCH', 'idx', '@n:[++inf 1]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[-+inf 1]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[+-inf 1]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[--inf 1]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[1 ++inf]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[1 -+inf]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[1 +-inf]').error()
+    env.expect('FT.SEARCH', 'idx', '@n:[1 --inf]').error()
+
+def testExplainNumericRange(env):
```
Comment (raz-mon) [inline tests/pytests/test_numbers.py:463]: Please add comments throughout the test, or an initial comment explaining the test.

File: tests/pytests/test_search_params.py [hunk 13]
```diff
@@ -324,16 +326,89 @@ def test_numeric_range(env):
...
+    env.expect('FT.SEARCH', 'idx', '@n:[(($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[1 (($n]', 'PARAMS', 2, 'n', 9).error()
+
+    # invalid syntax - signs before parenthesis are not allowed
+    env.expect('FT.SEARCH', 'idx', '@n:[+($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[-($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[-+($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[+-($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[--($n 9]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[++($n 9]', 'PARAMS', 2, 'n', 1).error()
```
Comment (raz-mon) [inline tests/pytests/test_search_params.py:411]: Why not allow this?..

**Review by raz-mon — COMMENTED** [review]
Nice!
See few comments

File: tests/pytests/test_numbers.py (hunk 12 above)
Comment (nafraf) [reply tests/pytests/test_numbers.py]: The tests done by `testExplainNumericRange(env)` were moved to `test:testExplain()`. [a448913](https://github.com/RediSearch/RediSearch/pull/4505/commits/a448913e7960c75d10bd6c2fcf6d8ce6e519dad7)

File: tests/pytests/test.py (hunk 10 above)
Comment (nafraf) [reply tests/pytests/test.py]: Fixed. Thanks.

File: tests/pytests/test.py (hunk 11 above)
Comment (nafraf) [reply tests/pytests/test.py]: No. Test removed.

File: tests/pytests/test_search_params.py [hunk 14]
```diff
@@ -292,7 +294,7 @@ def test_tags(env):
 
 
 def test_numeric_range(env):
-    env = Env(moduleArgs = 'DEFAULT_DIALECT 2')
+    env = Env(moduleArgs = 'DEFAULT_DIALECT 5')
```
Comment (raz-mon) [inline tests/pytests/test_search_params.py:297]: Consider adding a for loop over dialects 2 and 5 (as you did before) for the part that is common, and then adding another section with the part relevant only for dialect 5. This may help with clarifying the difference between them later on.

File: tests/pytests/test_search_params.py [hunk 15]
```diff
@@ -395,12 +453,38 @@ def test_numeric_range(env):
...
                'PARAMS', 2, 'param', 100).error()
-    env.expect('FT.SEARCH', 'idx', '@numval:[-$param]',
-               'PARAMS', 2, 'param', 100).error()
     env.expect('FT.SEARCH', 'idx', '@numval:[1  (($param]',
                'PARAMS', 2, 'param', 100).error()
     env.expect('FT.SEARCH', 'idx', '@numval:[(($param  1000]',
                'PARAMS', 2, 'param', 100).error()
+	# invalid syntax - multiple signs before parameters are not allowed
+    env.expect('FT.SEARCH', 'idx', '@n:[+-$n -+$n]', 'PARAMS', 2, 'n', 1).error()
+    env.expect('FT.SEARCH', 'idx', '@n:[++$n --$m]', 'PARAMS', 2, 'n', 1).error()
```
Comment (raz-mon) [inline tests/pytests/test_search_params.py:462]: Is there any special error that is raised here (I suspect no)?
You can add the `error().contains('...`)` function to verify the error thrown.

**Review by raz-mon — COMMENTED** [review]
Looks good.
See 2 small comments.

File: tests/pytests/test_search_params.py (hunk 14 above)
Comment (nafraf) [reply tests/pytests/test_search_params.py]: OK. Done. Thanks.

File: tests/pytests/test_search_params.py (hunk 15 above)
Comment (nafraf) [reply tests/pytests/test_search_params.py]: No. there is not an special error message, only `Syntax error`

**Review by oshadmi — APPROVED** [review]
